pkg/infrastructure/azure/dns.go

```diff
@@ -0,0 +1,242 @@
+	azureTags := make(map[string]*string)
+	for k, v := range in.InstallConfig.Config.Azure.UserTags {
+		azureTags[k] = ptr.To(v)
+	}
+	azureCluster := &capz.AzureCluster{}
+	key := client.ObjectKey{
+		Name:      in.InfraID,
+		Namespace: capiutils.Namespace,
+	}
+	if err := in.Client.Get(ctx, key, azureCluster); err != nil {
+		return fmt.Errorf("failed to get Azure cluster: %w", err)
+	}
+
+	ipIlb := azureCluster.Spec.NetworkSpec.APIServerLB.FrontendIPs[0].PrivateIPAddress
+
+	// TODO: Populate with IP after external LB is created.
+	var ipElb string
+
+	useIPv6 := false
+	for _, network := range in.InstallConfig.Config.Networking.ServiceNetwork {
+		if network.IP.To4() == nil {
+			useIPv6 = true
+		}
+	}
+
+	privateRecords := []recordPrivateList{}
+	ttl := int64(300)
+	recordType := arecord
+	if useIPv6 {
+		recordType = aaaarecord
+	}
+	publicRecords := createRecordSet(fmt.Sprintf("api.%s", in.InstallConfig.Config.ObjectMeta.Name), azureTags, ttl, recordType, ipElb, "")
+	privateRecords = append(privateRecords, createPrivateRecordSet("api-int", azureTags, ttl, recordType, ipIlb, ""))
+	privateRecords = append(privateRecords, createPrivateRecordSet("api", azureTags, ttl, recordType, ipIlb, ""))
+
+	// TODO: Populate with public LB FQDN. Placeholder text as value.
+	if !private {
+		cnameRecordName := apiExternalName
+		if useIPv6 {
+			cnameRecordName = apiExternalNameV6
+		}
+		publicRecords = createRecordSet(cnameRecordName, azureTags, ttl, cname, "", in.InstallConfig.Config.ClusterDomain())
+	}
+
+	session, err := in.InstallConfig.Azure.Session()
+	if err != nil {
+		return err
+	}
+	subscriptionID := session.Credentials.SubscriptionID
+	tokenCreds, err := azidentity.NewClientSecretCredential(session.Credentials.TenantID, session.Credentials.ClientID, session.Credentials.ClientSecret, nil)
+	if err != nil {
+		return err
+	}
+	recordSetClient, err := armdns.NewRecordSetsClient(subscriptionID, tokenCreds, nil)
+	if err != nil {
+		return err
+	}
+	privateRecordSetClient, err := armprivatedns.NewRecordSetsClient(subscriptionID, tokenCreds, nil)
+	if err != nil {
+		return err
+	}
+
+	logrus.Infof("zone:%s", zone)
+	logrus.Infof("private zone:%s", privatezone)
+	logrus.Infof("ip:%s", ipIlb)
+	logrus.Infof("%v", azureCluster)
+
+	// Create the records for api and api-int in the private zone and api.<clustername> for public zone.
+	// CAPI currently creates a record called "apiserver" instead of "api" so creating "api" for the installer in the private zone.
+	// TODO: check for internal publish and/or private cluster setup.
+	_, err = recordSetClient.CreateOrUpdate(ctx, baseDomainResourceGroup, zone, publicRecords.Name, publicRecords.RecordType, publicRecords.RecordSet, nil)
+	if err != nil {
+		return err
+	}
+
+	for _, record := range privateRecords {
+		_, err = privateRecordSetClient.CreateOrUpdate(ctx, resourceGroup, privatezone, record.RecordType, record.Name, record.RecordSet, nil)
+		if err != nil {
+			return err
```

Comment:
you should wrap all of the `return err`s in this file so that is easier to debug, e.g.:

```suggestion
			return fmt.Errorf("failed to create private record set %s: %w", record, err)
```

right now when there is an error in this function, we can see the error but we don't know where it came from:

```
time="2024-03-26T23:45:38Z" level=error msg="failed to fetch Cluster: failed to generate asset \"Cluster\": failed to create cluster: failed provisioning resources after infrastructure ready: <error message>
```